APE14.rst

```diff
@@ -0,0 +1,484 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Our proposal in this APE is the following:
+
+* To develop a common API that WCS objects can expose, either directly or
+  through thin wrappers, and which will allow functionality accessing the WCS
+  objects to be agnostic of the actual WCS representation. We want this API to
+  be as simple as possible and mainly return simple Python objects such as
+  strings, lists and arrays. We refer to this as the ‘low-level’ API.
+
+* To develop a high-level Astropy WCS object that can wrap any WCS object
+  providing the low-level API and be smarter about the kinds of objects to
+  accept or return to the user (including for example SkyCoord or Time objects).
+
+The idea of this two-tiered approach as opposed to a single-tier approach is
+that if we asked different WCS objects to provide a high-level API, this would
+(1) cause a lot of duplication of logic of constructing the appropriate Astropy
+objects, and (2) force those objects to return specifically Astropy objects,
+whereas having a package-independent API would be better.
+
+Low-level API
+^^^^^^^^^^^^^
+
+The proposed low-level API could be implemented as an abstract base class but we
+would not require all implementations to inherit from it (if they do not inherit
+from it, then they should still register with the ABC to allow code to check
+``isinstance(object, LowLevelWCSAPI))``. Existing WCS classes are then free to
+either implement that interface or provide thin wrappers providing this
+interface. A guiding philosophy of this interface is to use relatively primitive
+python objects (scalars, arrays, and strings), which can be “promoted” to more
+complex and useful Python objects in the high-level interface. This allows
+straightforward implementation of this interface even in C-implemented Python
+objects, without tying that into the high-level interface.
+
+Note that in this APE, when referring to arrays, we mean any Python object that
+follows the buffer protocol described in `PEP3118
+<https://www.python.org/dev/peps/pep-3118/>`_ rather than only specifically
+Numpy arrays (which do follow the buffer protocol). This allows for other array
+types, such as the `built-in Python array
+<https://docs.python.org/3/library/array.html>`_ objects or future advanced data
+structures
+
+The following class shows the required properties and methods the uniform
+low-level API recommends:
+
+.. code-block:: python
+
+    class LowLevelWCSAPI(metaclass=abc.ABCMeta):
+
+        @property
+        def n_dim_pixel(self):
+            """
+            The number of axes in the pixel coordinate system
+            """
+
+        @property
+        def n_dim_world(self):
+            """
+            The number of axes in the world coordinate system
+            """
+
+        @property
+        def shape_pixel(self):
+            """
+            The shape of the data that the WCS applies to as a tuple of
+            length ``n_dim_pixel`` (optional).
```

Comment:
Does this silently imply start of the data at 0?
I'm wondering about cases when, for some reason (overscan, reference pixels), the real data does not start from 0.  In this case, for exmaple, a distortion should be applied to ((xstart, xend), (ystart, yend)) and a tuple like this with xstart/ystart defaulting to 0 and xend/yend to NAXIS would be better.

Reply:
Are there use cases when the actual data size is needed?

Reply:
@nden - the shape can be useful for e.g. slicing of WCS objects, but i agree the bounds are useful too - can you review https://github.com/astrofrog/astropy-APEs/pull/15 and let me know if you are happy with those changes?